crates/bevy_ecs/src/query/error.rs

```
@@ -1,26 +1,93 @@
     AliasedMutability(Entity),
 }
 
+impl<'w> std::error::Error for QueryEntityError<'w> {}
+
+impl<'w> std::fmt::Display for QueryEntityError<'w> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::QueryDoesNotMatch(entity, world) => {
+                write!(
+                    f,
+                    "The query does not match the entity {entity}, which has components "
+                )?;
+                format_archetype(f, world, entity)
+            }
+            Self::NoSuchEntity(entity) => write!(f, "The entity {entity} does not exist"),
+            Self::AliasedMutability(entity) => write!(
+                f,
+                "The entity {entity} was requested mutably more than once"
+            ),
+        }
+    }
+}
+
+impl<'w> std::fmt::Debug for QueryEntityError<'w> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::QueryDoesNotMatch(entity, world) => {
+                write!(f, "QueryDoesNotMatch({entity} with components ")?;
+                format_archetype(f, world, entity)?;
+                write!(f, ")")
+            }
+            Self::NoSuchEntity(entity) => write!(f, "NoSuchEntity({entity})"),
+            Self::AliasedMutability(entity) => write!(f, "AliasedMutability({entity})"),
+        }
+    }
+}
+
+fn format_archetype(
+    f: &mut std::fmt::Formatter<'_>,
+    world: UnsafeWorldCell<'_>,
+    entity: Entity,
+) -> std::fmt::Result {
+    // We know entity is still life
+    let entity = world
+        .get_entity(entity)
+        .expect("entity does not belong to world");
+    for (i, component_id) in entity.archetype().components().enumerate() {
+        if i > 0 {
+            write!(f, ", ")?;
+        }
+        let name = world
+            .components()
+            .get_name(component_id)
+            .expect("entity does not belong to world");
```

If you expect it to exist, why not use `.name()`? If it might not exist, why not return an error?